### ISSUE TITLE
Declaring traittypes in Containers as uninstantiated classes

### ISSUE THREAD

**Issue by mpacer** [ipython/traitlets#457]
Feel free to ignore this, but I wasn't around when #51 was discussed… and this just came up in jupyter/nbconvert#695. So I wanted to weigh in on the question around deprecating a particular way of using uninstantiated traitlets.

There are two ways of using traitlets that were deprecated in 4.1 that were not documented in the migration docs http://traitlets.readthedocs.io/en/stable/migration.html. These use cases involve traitlets being used in slightly different ways. I think we should revisit the decision to deprecate one of them. 

Regardless of whether we continue to deprecate that case, the deprecations should be mentioned in the migration docs.

<details>
<summary>Click for more details about these two cases and why I think they are different.<br/><br/></summary>

1. traitlets that are actually instantiated on a class.
    ```python
    class MyClass():
    foo = Unicode() # ✓
    foo = Unicode # deprecated
    ```
2. traitlets defining which types are valid inside traitlets subclassed from `Container` (e.g., `List`).
    ```python
    class MyClass():
    foo = List(Unicode()) # ✓
    foo = List(Unicode) # deprecated
    ```

Deprecating 1. was absolutely the right call.  

I think that we should revisit the idea of deprecating 2.. 

<details>
<summary> Click here for my argument as to why we shouldn't deprecate use case 2. <br/><br/>
</summary>

In case 2, the actual use of a traitlet instance is somewhat strange from a type checking perspective. What you are effectively doing is declaring that the elements of this container need to be of the same _type_ as element that is listed. The only way I've ever seen type checking work involves declaring the _type_s that should be allowed, in python that would be the _class_es that should be allowed.

To follow the requirement that we only allow traitlet instances there would be analogous to a type checking syntax that declared that something should be a string by using literal strings. 

For example, the analogous case from the python type [annotations PEP][pytypes] was:

```python
def greeting(name: str) -> str:
    return 'Hello ' + name
```

but if using our convention it would be 

```python
def greeting(name: str()) -> str():
    return 'Hello ' + name
```

or equivalently:

```python
def greeting(name: '') -> '':
    return 'Hello ' + name
```

I realise that from an implementation point of view this introduces overhead because we also need to be able to put constraints on particular values, which means we can't only work on classes.
</details> 
     
**tldr**:; 
The use case in 2. is similar to type checking, and declaring types as uninstantiated classes is  a standard way of specifying types in signatures (including in [python type annotations][pytypes]).
</details>

[pytypes]: https://www.python.org/dev/peps/pep-0484/

**Comment by minrk** [comment]
I don't think the annotations PEP is applicable because traits are not the types for the actual values, they instantiate the *validators* of the values. TraitTypes are not useful in themselves, except for instantiating traits, which can then be used for validating actual values. So passing `Unicode` is implicitly asking traitlets to instantiate `Unicode()` for you, rather than instantiating it explicitly yourself. The PEP describes *actual* types, for which you can use `isinstance`. This is not the case for TraitTypes.

To draw a comparison: `str` is the right annotation because it provides something useful for validating (`isinstance(name, str)`). `str()` is not, because it does not provide something useful for validating (e.g. `isinstance(name, str())` doesn't work). The same goes for TraitType, but in reverse: `Unicode` does *not* provide something useful for validating (A TraitType is useless until instantiated), while `Unicode()` does (`Unicode().validate(obj, name)`).

So while I agree that `List(str)` would be preferable to `List(str())` (if that worked here), the same logic draws the reverse conclusion about whether to instantiate TraitTypes, since TraitTypes are not the types of the values, but only the constructors for the validators.

**Comment by mpacer** [comment]
OK, I think I follow. Closing!

**Comment by rmorshea** [comment]
@mpacer, any reason for reopening?

**Comment by rmorshea** [comment]
Closing in favor of a more descriptive issue: https://github.com/ipython/traitlets/issues/520
